{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97e7c3be",
   "metadata": {},
   "source": [
    "# Finding Nested Introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "129397db",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "from pybedtools import BedTool, Interval\n",
    "from tqdm import tqdm\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203baf1e",
   "metadata": {},
   "source": [
    "We define functions for loading different file types with SJ / intron data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c6033328",
   "metadata": {},
   "outputs": [],
   "source": [
    "SJ_FILE_COLUMN_NAMES = ['chromosome', 'start', 'end', 'strand', 'intron_motif',\n",
    "                        'annotated', 'reads_unique', 'reads_multimapped', 'max_overhang']\n",
    "BED_FILE_COLUMN_NAMES = ['chromosome', 'start', 'end', 'name', 'score', 'strand']\n",
    "\n",
    "\n",
    "def load_sj_dataframe(path: Path) -> pd.DataFrame:\n",
    "    df = pd.read_csv(path,\n",
    "                     delimiter='\\t',\n",
    "                     names=SJ_FILE_COLUMN_NAMES)\n",
    "    # STAR SJ files are 1-based and interval end is inclusive, while .bed files are 0-based and interval end is exclusive.\n",
    "    # We convert SJ files to .bed-style indexing (note that interval end is unchanged as 0-/1- indexing cancels out with exclusiveness / inclusiveness)\n",
    "    df['start'] -= 1\n",
    "\n",
    "    # We omit SJ with unspecified strand    \n",
    "    df = df[df['strand'] != 0]\n",
    "    df['strand'] = df['strand'].apply(lambda x: '+' if x == 1 else '-')\n",
    "\n",
    "    # We keep track how many samples support this SJ # this will be useful when aggregating dataframes from multiple samples\n",
    "    df['present_in_samples'] = 1\n",
    "    return df\n",
    "\n",
    "\n",
    "def load_bed_file_as_dataframe(file_path: Path) -> pd.DataFrame:\n",
    "    return pd.read_csv(file_path, delimiter='\\t', names=BED_FILE_COLUMN_NAMES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a34927aa",
   "metadata": {},
   "source": [
    "We load SJ files, filter them by some quality control, and aggregate them to a single dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "efc0aebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading SJ files: 16it [00:01,  8.67it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>reads_unique</th>\n",
       "      <th>reads_multimapped</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>present_in_samples</th>\n",
       "      <th>present_in_samples_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456211.1</td>\n",
       "      <td>54472</td>\n",
       "      <td>55342</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL456211.1</td>\n",
       "      <td>55514</td>\n",
       "      <td>61021</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456211.1</td>\n",
       "      <td>167988</td>\n",
       "      <td>175871</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>9</td>\n",
       "      <td>0.5625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL456211.1</td>\n",
       "      <td>176034</td>\n",
       "      <td>176916</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "      <td>52</td>\n",
       "      <td>50</td>\n",
       "      <td>5</td>\n",
       "      <td>0.3125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL456211.1</td>\n",
       "      <td>177088</td>\n",
       "      <td>182218</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156901</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90760614</td>\n",
       "      <td>90762078</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156902</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90785979</td>\n",
       "      <td>90793295</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>273</td>\n",
       "      <td>1479</td>\n",
       "      <td>50</td>\n",
       "      <td>15</td>\n",
       "      <td>0.9375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156903</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90793417</td>\n",
       "      <td>90816348</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>884</td>\n",
       "      <td>36</td>\n",
       "      <td>50</td>\n",
       "      <td>15</td>\n",
       "      <td>0.9375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156904</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90793680</td>\n",
       "      <td>90816348</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>15</td>\n",
       "      <td>50</td>\n",
       "      <td>15</td>\n",
       "      <td>0.9375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156905</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90800534</td>\n",
       "      <td>90807569</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156906 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome     start       end strand  intron_motif  annotated  \\\n",
       "0       GL456211.1     54472     55342      +             1          0   \n",
       "1       GL456211.1     55514     61021      +             1          0   \n",
       "2       GL456211.1    167988    175871      +             1          1   \n",
       "3       GL456211.1    176034    176916      +             1          1   \n",
       "4       GL456211.1    177088    182218      +             1          1   \n",
       "...            ...       ...       ...    ...           ...        ...   \n",
       "156901        chrY  90760614  90762078      +             1          1   \n",
       "156902        chrY  90785979  90793295      +             1          1   \n",
       "156903        chrY  90793417  90816348      +             1          1   \n",
       "156904        chrY  90793680  90816348      +             1          1   \n",
       "156905        chrY  90800534  90807569      +             1          0   \n",
       "\n",
       "        reads_unique  reads_multimapped  max_overhang  present_in_samples  \\\n",
       "0                  5                  0            41                   1   \n",
       "1                  7                  0            29                   1   \n",
       "2                 89                 10            50                   9   \n",
       "3                 37                 52            50                   5   \n",
       "4                  5                  1            40                   1   \n",
       "...              ...                ...           ...                 ...   \n",
       "156901            15                  2            46                   1   \n",
       "156902           273               1479            50                  15   \n",
       "156903           884                 36            50                  15   \n",
       "156904           187                 15            50                  15   \n",
       "156905            12                  4            50                   2   \n",
       "\n",
       "        present_in_samples_fraction  \n",
       "0                            0.0625  \n",
       "1                            0.0625  \n",
       "2                            0.5625  \n",
       "3                            0.3125  \n",
       "4                            0.0625  \n",
       "...                             ...  \n",
       "156901                       0.0625  \n",
       "156902                       0.9375  \n",
       "156903                       0.9375  \n",
       "156904                       0.9375  \n",
       "156905                       0.1250  \n",
       "\n",
       "[156906 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_folder_path = Path('./splice_junction_files')\n",
    "sj_dataframes = [load_sj_dataframe(path) for path in tqdm(sj_folder_path.iterdir(), desc='Loading SJ files')]\n",
    "\n",
    "\n",
    "def quality_control_sj(df: pd.DataFrame, min_max_overhang=20, min_unique_reads=5) -> pd.DataFrame:\n",
    "    df = df[df['max_overhang'] >= min_max_overhang]\n",
    "    df = df[df['reads_unique'] >= min_unique_reads]\n",
    "    return df\n",
    "\n",
    "\n",
    "sj_dataframes = [quality_control_sj(sj) for sj in sj_dataframes]\n",
    "\n",
    "sj_aggregated = pd.concat(sj_dataframes).groupby(['chromosome', 'start', 'end',\n",
    "                                                  'strand', 'intron_motif', 'annotated']).agg(\n",
    "    {'reads_unique': 'sum',\n",
    "     'reads_multimapped': 'sum',\n",
    "     'max_overhang': 'max',\n",
    "     'present_in_samples': 'sum'}).reset_index()\n",
    "\n",
    "sj_aggregated['present_in_samples_fraction'] = sj_aggregated['present_in_samples'] / len(sj_dataframes)\n",
    "sj_aggregated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7c648b",
   "metadata": {},
   "source": [
    "We may further e.g. filter only SJ present in all samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1af0f45a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>reads_unique</th>\n",
       "      <th>reads_multimapped</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>present_in_samples</th>\n",
       "      <th>present_in_samples_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>16407</td>\n",
       "      <td>30698</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>961</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>30806</td>\n",
       "      <td>31058</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1138</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>31127</td>\n",
       "      <td>31369</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>661</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>31471</td>\n",
       "      <td>34137</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1280</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>34345</td>\n",
       "      <td>34623</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2314</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156882</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169313022</td>\n",
       "      <td>169313587</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>837</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156883</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169313674</td>\n",
       "      <td>169315601</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>520</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156886</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169317147</td>\n",
       "      <td>169319437</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>606</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156887</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169319590</td>\n",
       "      <td>169319975</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156888</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169319590</td>\n",
       "      <td>169320114</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>279</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81774 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      start        end strand  intron_motif  annotated  \\\n",
       "13      GL456216.1      16407      30698      +             1          1   \n",
       "16      GL456216.1      30806      31058      +             1          1   \n",
       "17      GL456216.1      31127      31369      +             1          1   \n",
       "19      GL456216.1      31471      34137      +             1          1   \n",
       "20      GL456216.1      34345      34623      +             1          1   \n",
       "...            ...        ...        ...    ...           ...        ...   \n",
       "156882        chrX  169313022  169313587      -             2          1   \n",
       "156883        chrX  169313674  169315601      -             2          1   \n",
       "156886        chrX  169317147  169319437      -             2          1   \n",
       "156887        chrX  169319590  169319975      -             2          1   \n",
       "156888        chrX  169319590  169320114      -             2          1   \n",
       "\n",
       "        reads_unique  reads_multimapped  max_overhang  present_in_samples  \\\n",
       "13               961                  0            50                  16   \n",
       "16              1138                  1            50                  16   \n",
       "17               661                  0            50                  16   \n",
       "19              1280                  0            50                  16   \n",
       "20              2314                  0            50                  16   \n",
       "...              ...                ...           ...                 ...   \n",
       "156882           837                  0            50                  16   \n",
       "156883           520                  0            50                  16   \n",
       "156886           606                  1            50                  16   \n",
       "156887           295                  0            50                  16   \n",
       "156888           279                  0            50                  16   \n",
       "\n",
       "        present_in_samples_fraction  \n",
       "13                              1.0  \n",
       "16                              1.0  \n",
       "17                              1.0  \n",
       "19                              1.0  \n",
       "20                              1.0  \n",
       "...                             ...  \n",
       "156882                          1.0  \n",
       "156883                          1.0  \n",
       "156886                          1.0  \n",
       "156887                          1.0  \n",
       "156888                          1.0  \n",
       "\n",
       "[81774 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_aggregated = sj_aggregated[sj_aggregated['present_in_samples_fraction'] == 1]\n",
    "sj_aggregated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f94028f4",
   "metadata": {},
   "source": [
    "We define function for finding nested introns and apply it to dataframe with SJ:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4af25d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_nested_introns(df_1: pd.DataFrame, df_2: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Find nested introns (SJ). The potential 'outside' and 'nested' introns are specified by different arguments (\n",
    "      df_1 resp. df_2), which may be useful if we wish to apply different pre-filtering logic on them.\n",
    "    :param df_1: Dataframe with potential 'outside' introns.\n",
    "    :param df_2: Dataframe with potential 'nested' introns.\n",
    "    :return: Dataframe same as df_1, with added columns 'has_nested_intron' and 'nested_intron_indices'. The column\n",
    "    'has_nested_intron' contain boolean flag specifying whether this intron contains any nested introns. The column\n",
    "    'nested_intron_indices' contain indices (from df_2) of nested introns.\n",
    "    \"\"\"\n",
    "    df_output = df_1.copy()\n",
    "\n",
    "    bedtool_1 = BedTool([Interval(chrom=row['chromosome'],\n",
    "                                  start=row['start'],\n",
    "                                  end=row['end'],\n",
    "                                  name=index,\n",
    "                                  strand=row['strand']) for index, row in df_1.iterrows()]).sort()\n",
    "\n",
    "    bedtool_2 = BedTool([Interval(chrom=row['chromosome'],\n",
    "                                  start=row['start'],\n",
    "                                  end=row['end'],\n",
    "                                  name=index,\n",
    "                                  strand=row['strand']) for index, row in df_2.iterrows()]).sort()\n",
    "\n",
    "    intersection_bedtool = bedtool_1.intersect(bedtool_2, sorted=True, s=True, wa=True, wb=True)\n",
    "    intersections_by_query_id = defaultdict(list)\n",
    "    for intersection_interval in intersection_bedtool:\n",
    "        intersections_by_query_id[int(intersection_interval.fields[3])].append(int(intersection_interval.fields[9]))\n",
    "\n",
    "    has_nested_intron: list[bool] = []\n",
    "    nested_intron_indices: list[list[int]] = []\n",
    "\n",
    "    for index, row in tqdm(df_1.iterrows(), desc='Finding nested introns', total=len(df_output)):\n",
    "        intersections_indices = intersections_by_query_id.get(index)\n",
    "        if not intersections_indices:\n",
    "            has_nested_intron.append(False)\n",
    "            nested_intron_indices.append([])\n",
    "            continue\n",
    "\n",
    "        intersections_df = df_2.loc[intersections_indices]\n",
    "        nested_introns_df = intersections_df[intersections_df['start'] > row['start']]\n",
    "        nested_introns_df = nested_introns_df[nested_introns_df['end'] < row['end']]\n",
    "\n",
    "        nested_introns_indices = list(nested_introns_df.index)\n",
    "        has_nested_intron.append(bool(nested_introns_indices))\n",
    "        nested_intron_indices.append(nested_introns_indices)\n",
    "\n",
    "    df_output['has_nested_intron'] = has_nested_intron\n",
    "    df_output['nested_intron_indices'] = nested_intron_indices\n",
    "    return df_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d2a80ab7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.gf0xh1r0.tmp has inconsistent naming convention for record:\n",
      "chr1\t4774516\t4777524\t65\t.\t-\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.gf0xh1r0.tmp has inconsistent naming convention for record:\n",
      "chr1\t4774516\t4777524\t65\t.\t-\n",
      "\n",
      "Finding nested introns: 100%|█████████████████████████████████████| 81774/81774 [02:01<00:00, 674.43it/s]\n"
     ]
    }
   ],
   "source": [
    "sj_aggregated = find_nested_introns(df_1=sj_aggregated, df_2=sj_aggregated)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98371074",
   "metadata": {},
   "source": [
    "Some SJ have nested SJ inside them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "751717bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>reads_unique</th>\n",
       "      <th>reads_multimapped</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>present_in_samples</th>\n",
       "      <th>present_in_samples_fraction</th>\n",
       "      <th>has_nested_intron</th>\n",
       "      <th>nested_intron_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63170914</td>\n",
       "      <td>63176585</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1608</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[2625]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3730</th>\n",
       "      <td>chr1</td>\n",
       "      <td>82735507</td>\n",
       "      <td>82750519</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1005</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[3732]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4214</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88071678</td>\n",
       "      <td>88215043</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2639</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[4219]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4256</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88268026</td>\n",
       "      <td>88275041</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>956</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[4257, 4258]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7260</th>\n",
       "      <td>chr1</td>\n",
       "      <td>156041002</td>\n",
       "      <td>156059410</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>349</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[7261]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149076</th>\n",
       "      <td>chr9</td>\n",
       "      <td>83925359</td>\n",
       "      <td>83988775</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>501</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[149082]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150003</th>\n",
       "      <td>chr9</td>\n",
       "      <td>103217518</td>\n",
       "      <td>103219801</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>474</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[150004]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150633</th>\n",
       "      <td>chr9</td>\n",
       "      <td>107659244</td>\n",
       "      <td>107668852</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1227</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[150637, 150642]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151026</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108517240</td>\n",
       "      <td>108528588</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>469</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[151030]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152964</th>\n",
       "      <td>chrX</td>\n",
       "      <td>13053350</td>\n",
       "      <td>13093923</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>247</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[152967]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>118 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome      start        end strand  intron_motif  annotated  \\\n",
       "2623         chr1   63170914   63176585      -             2          1   \n",
       "3730         chr1   82735507   82750519      +             1          1   \n",
       "4214         chr1   88071678   88215043      +             1          1   \n",
       "4256         chr1   88268026   88275041      -             2          1   \n",
       "7260         chr1  156041002  156059410      +             1          1   \n",
       "...           ...        ...        ...    ...           ...        ...   \n",
       "149076       chr9   83925359   83988775      +             1          0   \n",
       "150003       chr9  103217518  103219801      -             2          1   \n",
       "150633       chr9  107659244  107668852      -             2          1   \n",
       "151026       chr9  108517240  108528588      +             1          1   \n",
       "152964       chrX   13053350   13093923      +             1          0   \n",
       "\n",
       "        reads_unique  reads_multimapped  max_overhang  present_in_samples  \\\n",
       "2623            1608                  1            50                  16   \n",
       "3730            1005                  0            50                  16   \n",
       "4214            2639                  1            50                  16   \n",
       "4256             956                  0            50                  16   \n",
       "7260             349                  2            50                  16   \n",
       "...              ...                ...           ...                 ...   \n",
       "149076           501                  0            50                  16   \n",
       "150003           474                  0            50                  16   \n",
       "150633          1227                  0            50                  16   \n",
       "151026           469                  0            50                  16   \n",
       "152964           247                  0            50                  16   \n",
       "\n",
       "        present_in_samples_fraction  has_nested_intron nested_intron_indices  \n",
       "2623                            1.0               True                [2625]  \n",
       "3730                            1.0               True                [3732]  \n",
       "4214                            1.0               True                [4219]  \n",
       "4256                            1.0               True          [4257, 4258]  \n",
       "7260                            1.0               True                [7261]  \n",
       "...                             ...                ...                   ...  \n",
       "149076                          1.0               True              [149082]  \n",
       "150003                          1.0               True              [150004]  \n",
       "150633                          1.0               True      [150637, 150642]  \n",
       "151026                          1.0               True              [151030]  \n",
       "152964                          1.0               True              [152967]  \n",
       "\n",
       "[118 rows x 13 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_with_nested_introns = sj_aggregated[sj_aggregated['has_nested_intron']]\n",
    "sj_with_nested_introns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26275f25",
   "metadata": {},
   "source": [
    "The column 'nested_intron_indices' keeps indices (of second dataframe passed to find_nested_introns() function) which we may use to retrieve the nested SJ:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d488c0c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Example SJ: \n",
      "chromosome                             chr1\n",
      "start                              88268026\n",
      "end                                88275041\n",
      "strand                                    -\n",
      "intron_motif                              2\n",
      "annotated                                 1\n",
      "reads_unique                            956\n",
      "reads_multimapped                         0\n",
      "max_overhang                             50\n",
      "present_in_samples                       16\n",
      "present_in_samples_fraction             1.0\n",
      "has_nested_intron                      True\n",
      "nested_intron_indices          [4257, 4258]\n",
      "Name: 4256, dtype: object\n",
      "--------------------------------------------------\n",
      "Example nested introns of the SJ above:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>reads_unique</th>\n",
       "      <th>reads_multimapped</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>present_in_samples</th>\n",
       "      <th>present_in_samples_fraction</th>\n",
       "      <th>has_nested_intron</th>\n",
       "      <th>nested_intron_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4257</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88269413</td>\n",
       "      <td>88270201</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1315</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4258</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88270299</td>\n",
       "      <td>88272728</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>800</td>\n",
       "      <td>2</td>\n",
       "      <td>43</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome     start       end strand  intron_motif  annotated  \\\n",
       "4257       chr1  88269413  88270201      -             2          1   \n",
       "4258       chr1  88270299  88272728      -             2          1   \n",
       "\n",
       "      reads_unique  reads_multimapped  max_overhang  present_in_samples  \\\n",
       "4257          1315                  1            50                  16   \n",
       "4258           800                  2            43                  16   \n",
       "\n",
       "      present_in_samples_fraction  has_nested_intron nested_intron_indices  \n",
       "4257                          1.0              False                    []  \n",
       "4258                          1.0              False                    []  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_sj = sj_with_nested_introns.iloc[3]\n",
    "example_nested_introns = sj_aggregated.loc[example_sj['nested_intron_indices']]\n",
    "print(f\"Example SJ: \\n{example_sj}\")\n",
    "print(50 * '-')\n",
    "print(\"Example nested introns of the SJ above:\")\n",
    "example_nested_introns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a69d62",
   "metadata": {},
   "source": [
    "We may find nested SJ inside different reference introns, if we wish to apply different filtering logic on the nested and 'outside' introns.\n",
    "We may e.g. find SJ that are nested inside introns from the reference genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "babc787b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.4cai5zxo.tmp has inconsistent naming convention for record:\n",
      "chr1\t3207317\t3213438\t192\t.\t-\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.4cai5zxo.tmp has inconsistent naming convention for record:\n",
      "chr1\t3207317\t3213438\t192\t.\t-\n",
      "\n",
      "Finding nested introns: 100%|██████████████████████████████████| 225163/225163 [01:50<00:00, 2039.14it/s]\n"
     ]
    }
   ],
   "source": [
    "reference_introns = load_bed_file_as_dataframe('./reference_genome/introns.bed')\n",
    "reference_introns = find_nested_introns(reference_introns, sj_aggregated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "813c7222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>has_nested_intron</th>\n",
       "      <th>nested_intron_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15838</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5069063</td>\n",
       "      <td>5336948</td>\n",
       "      <td>ENSMUSG00000096054.2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>[10046, 10047, 10048, 10049, 10050, 10051, 100...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48034</th>\n",
       "      <td>chr12</td>\n",
       "      <td>104731987</td>\n",
       "      <td>104751747</td>\n",
       "      <td>ENSMUSG00000041415.9</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>[34917]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50423</th>\n",
       "      <td>chr13</td>\n",
       "      <td>17672602</td>\n",
       "      <td>17694638</td>\n",
       "      <td>ENSMUSG00000055137.6</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>[36699]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76176</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23180009</td>\n",
       "      <td>23218762</td>\n",
       "      <td>ENSMUSG00000068463.3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>[55715, 55717, 55720, 55724]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79182</th>\n",
       "      <td>chr16</td>\n",
       "      <td>85030365</td>\n",
       "      <td>85056323</td>\n",
       "      <td>ENSMUSG00000022892.10</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>[57891]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91432</th>\n",
       "      <td>chr18</td>\n",
       "      <td>32135951</td>\n",
       "      <td>32139509</td>\n",
       "      <td>ENSMUSG00000024386.8</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>[67593, 67595]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163394</th>\n",
       "      <td>chr6</td>\n",
       "      <td>40686610</td>\n",
       "      <td>40714600</td>\n",
       "      <td>ENSMUSG00000068587.9</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>[118413, 118414, 118415, 118416, 118417, 11841...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163395</th>\n",
       "      <td>chr6</td>\n",
       "      <td>40714729</td>\n",
       "      <td>40728951</td>\n",
       "      <td>ENSMUSG00000068587.9</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>[118437, 118439, 118440, 118442, 118443, 118444]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175880</th>\n",
       "      <td>chr7</td>\n",
       "      <td>26173604</td>\n",
       "      <td>26186580</td>\n",
       "      <td>ENSMUSG00000040660.6</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>[126381]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome      start        end                   name score strand  \\\n",
       "15838       chr10    5069063    5336948   ENSMUSG00000096054.2     .      -   \n",
       "48034       chr12  104731987  104751747   ENSMUSG00000041415.9     .      -   \n",
       "50423       chr13   17672602   17694638   ENSMUSG00000055137.6     .      -   \n",
       "76176       chr16   23180009   23218762   ENSMUSG00000068463.3     .      +   \n",
       "79182       chr16   85030365   85056323  ENSMUSG00000022892.10     .      -   \n",
       "91432       chr18   32135951   32139509   ENSMUSG00000024386.8     .      -   \n",
       "163394       chr6   40686610   40714600   ENSMUSG00000068587.9     .      +   \n",
       "163395       chr6   40714729   40728951   ENSMUSG00000068587.9     .      +   \n",
       "175880       chr7   26173604   26186580   ENSMUSG00000040660.6     .      +   \n",
       "\n",
       "        has_nested_intron                              nested_intron_indices  \n",
       "15838                True  [10046, 10047, 10048, 10049, 10050, 10051, 100...  \n",
       "48034                True                                            [34917]  \n",
       "50423                True                                            [36699]  \n",
       "76176                True                       [55715, 55717, 55720, 55724]  \n",
       "79182                True                                            [57891]  \n",
       "91432                True                                     [67593, 67595]  \n",
       "163394               True  [118413, 118414, 118415, 118416, 118417, 11841...  \n",
       "163395               True   [118437, 118439, 118440, 118442, 118443, 118444]  \n",
       "175880               True                                           [126381]  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_introns_with_nested_sj = reference_introns[reference_introns['has_nested_intron']]\n",
    "reference_introns_with_nested_sj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "140d9757",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Example intron with nested SJ: \n",
      "chromosome                                      chr16\n",
      "start                                        23180009\n",
      "end                                          23218762\n",
      "name                             ENSMUSG00000068463.3\n",
      "score                                               .\n",
      "strand                                              +\n",
      "has_nested_intron                                True\n",
      "nested_intron_indices    [55715, 55717, 55720, 55724]\n",
      "Name: 76176, dtype: object\n",
      "--------------------------------------------------\n",
      "Nested SJ of the intron above:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>intron_motif</th>\n",
       "      <th>annotated</th>\n",
       "      <th>reads_unique</th>\n",
       "      <th>reads_multimapped</th>\n",
       "      <th>max_overhang</th>\n",
       "      <th>present_in_samples</th>\n",
       "      <th>present_in_samples_fraction</th>\n",
       "      <th>has_nested_intron</th>\n",
       "      <th>nested_intron_indices</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55715</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23192261</td>\n",
       "      <td>23192982</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1841</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55717</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23193143</td>\n",
       "      <td>23193498</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1064</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55720</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23193567</td>\n",
       "      <td>23200985</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>792</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55724</th>\n",
       "      <td>chr16</td>\n",
       "      <td>23201214</td>\n",
       "      <td>23203490</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>893</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome     start       end strand  intron_motif  annotated  \\\n",
       "55715      chr16  23192261  23192982      +             1          0   \n",
       "55717      chr16  23193143  23193498      +             1          0   \n",
       "55720      chr16  23193567  23200985      +             1          0   \n",
       "55724      chr16  23201214  23203490      +             1          0   \n",
       "\n",
       "       reads_unique  reads_multimapped  max_overhang  present_in_samples  \\\n",
       "55715          1841                  0            50                  16   \n",
       "55717          1064                  0            50                  16   \n",
       "55720           792                  0            50                  16   \n",
       "55724           893                  0            50                  16   \n",
       "\n",
       "       present_in_samples_fraction  has_nested_intron nested_intron_indices  \n",
       "55715                          1.0              False                    []  \n",
       "55717                          1.0              False                    []  \n",
       "55720                          1.0              False                    []  \n",
       "55724                          1.0              False                    []  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_intron = reference_introns_with_nested_sj.iloc[3]\n",
    "example_nested_sj = sj_aggregated.loc[example_intron['nested_intron_indices']]\n",
    "print(f\"Example intron with nested SJ: \\n{example_intron}\")\n",
    "print(50 * '-')\n",
    "print(\"Nested SJ of the intron above:\")\n",
    "example_nested_sj"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
